src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php

```diff
@@ -128,26 +128,59 @@ public function onKernelResponse(ResponseEvent $event): void
      */
     protected function injectToolbar(Response $response, Request $request, array $nonces): void
     {
-        $content = $response->getContent();
-        $pos = strripos($content, '</body>');
-
-        if (false !== $pos) {
-            $toolbar = "\n".str_replace("\n", '', $this->twig->render(
-                '@WebProfiler/Profiler/toolbar_js.html.twig',
-                [
-                    'full_stack' => class_exists(FullStack::class),
-                    'excluded_ajax_paths' => $this->excludedAjaxPaths,
-                    'token' => $response->headers->get('X-Debug-Token'),
-                    'request' => $request,
-                    'csp_script_nonce' => $nonces['csp_script_nonce'] ?? null,
-                    'csp_style_nonce' => $nonces['csp_style_nonce'] ?? null,
-                ]
-            ))."\n";
-            $content = substr($content, 0, $pos).$toolbar.substr($content, $pos);
-            $response->setContent($content);
+        if ($response instanceof StreamedResponse) {
+            $callback = $response->getCallback();
+            if (false !== strripos($response->headers->get('Content-Type'), 'text/html')) {
+                $toolbarHTMLContent = $this->getToolbarHTML($request, $response->headers->get('X-Debug-Token'), $nonces);
+                $injectedCallback = static function () use ($toolbarHTMLContent, $callback): void {
+                    ob_start(function (string $buffer, int $phase) use ($toolbarHTMLContent): string {
+                        $pos = strripos($buffer, '</body>');
+                        if (false !== $pos) {
+                            $buffer = substr($buffer, 0, $pos).$toolbarHTMLContent.substr($buffer, $pos);
+                        }
+
+                        return $buffer;
+                    }, 8); // length of '</body>'
+
+                    ($callback)();
+                    ob_end_flush();
+                };
+                $response->setCallback($injectedCallback);
+            }
+        } else {
+            $content = $response->getContent();
+            $pos = strripos($content, '</body>');
+
+            if (false !== $pos) {
+                $response->setContent(
+                    $this->renderToolbarInContent($content, $pos, $response->headers->get('X-Debug-Token'), $request, $nonces)
+                );
+            }
         }
     }
 
+    protected function renderToolbarInContent(string $content, int $pos, ?string $debugToken, Request $request, array $nonces): string
```

Comment:
why adding a new protected method ?

Reply:
You're right, only used at one place. Removed.